### ISSUE TITLE
Shadow can take a long time to start processes

### ISSUE THREAD

**Issue by stevenengler** [shadow/shadow#2910]
This has been an issue in Shadow for a while, but it's become much more noticeable when running on our new benchmark runner. In some simulations, Shadow can take a very long time to start processes. This is not deterministic and only happens occasionally.

In the graph below you can see that Shadow seems to get stuck for a while around the simtimes 0:00, 4:00, and 5:00. These simtimes align with when processes are started in the Shadow config (with hundreds of hosts; shortened to only show start times):

```yaml
  4uthority1:
    processes:
    - path: ~/.local/bin/tor
      start_time: 1 # 0:01
    - path: ~/.local/bin/oniontrace
      start_time: 2 # 0:02
  server1exit:
    processes:
    - path: ~/.local/bin/tgen
      start_time: 300 # 5:00
  perfclient1exit:
    processes:
    - path: ~/.local/bin/tor
      start_time: 240 # 4:00
    - path: ~/.local/bin/oniontrace
      start_time: 241 # 4:01
    - path: ~/.local/bin/tgen
      start_time: 300 # 5:00
```

Here we see Shadow taking over an hour (about 33% of the total simulation runtime) to launch processes.

https://github.com/shadow/benchmark-results/tree/master/tor/2023-05-07-T19-52-20

![1683554926_grim](https://user-images.githubusercontent.com/3708797/236846672-ed0b5661-9b93-4db6-97d2-2a0e7b538ea0.png)

I'm not really sure why this might be happening, or what we could do to debug it as it's not reproducible. It could be the fork itself (`clone` in `posix_spawn`) that takes a while, but we use `CLONE_VM` which removes the obvious case of expensive page-table duplication. Maybe there could be some resource exhaustion when starting multiple processes that slows down the simulation. For example does Linux re-seed anything when forking processes (maybe something along the lines of ["Random numbers and virtual-machine forks"](https://lwn.net/Articles/887207/)), which could potentially use and deplete system entropy. Or maybe a library used by the managed processes uses system entropy to initialize something in a `__constructor__` that runs before Shadow can intercept syscalls? Openssl could be a contender for this, but we see the delay even when launching tgen, so this seems unlikely.

**Comment by stevenengler** [comment]
It doesn't seem like it is the fork, since in the log we see for example:

```
00:05:43.448849 [37098:shadow-worker] 00:00:05.000000000 [INFO] [relay153guard:109.202.212.1] [process.rs:812] [shadow_rs::host::process] starting process 'oniontrace'
00:05:43.448864 [37098:shadow-worker] 00:00:05.000000000 [INFO] [relay153guard:109.202.212.1] [managed_thread.rs:139] [shadow_rs::host::managed_thread] forking new mthread with environment '["LD_PRELOAD=/__w/benchmark/benchmark/shadow/build/src/lib/injector_preload/libshadow_injector.so:/__w/benchmark/benchmark/shadow/build/src/lib/libc_preload/libshadow_libc.so:/__w/benchmark/benchmark/shadow/build/src/lib/openssl_preload/libshadow_openssl_rng.so:/__w/benchmark/benchmark/shadow/build/src/lib/openssl_preload/libshadow_openssl_crypto.so", "SHADOW_LOG_LEVEL=3", "SHADOW_LOG_START_TIME=1683491272662007", "SHADOW_SHM_THREAD_BLK=761864,1048576,4328,/shadow_shmemfile_276296.436476266-36972", "SHADOW_IPC_BLK=722952,1048576,592,/shadow_shmemfile_276296.436476266-36972"]', arguments '["/github/home/.local/bin/oniontrace", "Mode=log", "TorControlPort=9051", "LogLevel=info", "Events=BW,CIRC"]', and working directory '"/tmp/sim-results/tornet-0.05-1/shadow.data/hosts/relay153guard"'
00:05:43.449466 [37098:shadow-worker] 00:00:05.000000000 [INFO] [relay153guard:109.202.212.1] [process.rs:834] [shadow_rs::host::process] process 'oniontrace' started
00:05:43.567622 [36972:shadow] n/a [INFO] [n/a] [manager.rs:753] [shadow_rs::core::manager] Process resource usage at simtime 5006392000 reported by getrusage(): ru_maxrss=2.069 GiB, ru_utime=0.500 minutes, ru_stime=0.117 minutes, ru_nvcsw=503854, ru_nivcsw=25265
00:05:43.567633 [36972:shadow] n/a [INFO] [n/a] [manager.rs:771] [shadow_rs::core::manager] System memory usage in bytes at simtime 5006392000 ns reported by /proc/meminfo: {"mem_total":540632530944,"mem_free":520356839424,"swap_total":1023406080,"swap_free":1023406080,"buffers":1442594816,"cached":14154764288,"s_reclaimable":867012608,"shmem":344399872}
00:16:10.782593 [37102:shadow-worker] 00:00:06.000000000 [INFO] [relay248middle:185.236.203.203] [process.rs:812] [shadow_rs::host::process] starting process 'oniontrace'
```

The real time jumps from 5:43 to 16:10 after the oniontrace process had started (and the sim time changes from 0:05 to 0:06).

**Comment by stevenengler** [comment]
Since upgrading the benchmark machine (debian 11 to debian 12, upgraded kernel, etc), this behaviour seems to have gone away. It's not clear if it went away because of the software upgrades, or maybe because the machine was restarted. If it was the latter, the behaviour may resurface at some point.

**Comment by robgjansen** [comment]
I made several improvements while upgrading the benchmark runner to Debian 12. The two changes that I believe helped to fix this problem were:

- I moved the working directory that the container uses during the simulations from a low-speed hard disk drive mount to a high-speed NVME mount. The archive dir is still on the HDD mount, but that mount is only used after the sim to copy out the results. During the sim, IO operations should all be high speed now.
- The [preferred Docker storage driver is `overlay2`](https://docs.docker.com/storage/storagedriver/select-storage-driver/), which IIRC has kernel support. In Debian 11, the `overlay2` driver failed and I had to fall back to the `fuse-overlayfs` driver. In Debian 12, I was able to get `overlay2` working.

I was hopeful these changes would fix the problem, and I'm glad we haven't noticed the issue again since :) Do we have enough evidence it's working correctly now to close this issue?
